extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryFrameProcessor.java

```diff
@@ -216,32 +217,58 @@ Current approach with R&C and operators materialize a single R&C for processing.
      Most of the window operations like SUM(), RANK(), RANGE() etc. can be made with 2 passes of the data. We might think to reimplement them in the MSQ way so that we do not have to materialize so much data.
      */
 
+    // If there are rows pending flush, flush them and run again before processing any more rows.
+    if (frameHasRowsPendingFlush()) {
+      flushAllRowsAndCols(resultRowAndCols, rowId);
+      return ReturnOrAwait.runAgain();
+    }
+
     if (partitionColumnNames.isEmpty()) {
       // Scenario 1: Query has atleast one window function with an OVER() clause without a PARTITION BY.
       if (inputChannel.canRead()) {
         final Frame frame = inputChannel.read();
         convertRowFrameToRowsAndColumns(frame);
         return ReturnOrAwait.runAgain();
-      } else if (inputChannel.isFinished()) {
-        runAllOpsOnMultipleRac(frameRowsAndCols);
+      }
+
+      if (inputChannel.isFinished()) {
+        // If no rows are flushed yet, process all rows.
+        if (rowId.get() == 0) {
+          runAllOpsOnMultipleRac(frameRowsAndCols);
+        }
+
+        // If there are still rows pending after operations, run again.
+        if (frameHasRowsPendingFlush()) {
+          return ReturnOrAwait.runAgain();
+        }
         return ReturnOrAwait.returnObject(Unit.instance());
-      } else {
-        return ReturnOrAwait.awaitAll(inputChannels().size());
       }
+      return ReturnOrAwait.awaitAll(inputChannels().size());
     }
 
+    // This helps us avoid OOM issues, as it ensures that we don't keep accumulating the processed rows in-memory, if
+    // the rate of processing of the rows is higher than the allowed capacity of frame writer in each iteration of runIncrementally.
+    clearRACBuffers();
```

Comment:
I don't think there is a need for this comment; the normal operation should not lead to OOM; that should be designed for.
are there any such cases which are known?

in general I feel like this doesn't make much sense:
this method sets `rowId` to `0` - if that would happen when we do have something still to write; that will make it loose data...

why can't the same be done when the last part of the frame was written?
call it at the end of `writeRacToFrame` ?

Reply:
> I don't think there is a need for this comment; the normal operation should not lead to OOM; that should be designed for.
are there any such cases which are known?

If we run the following query that deals with around 50M rows:
```sql
select trip_id, row_number() over(partition by trip_id) as c1 from "trips_xaa" where __time < TIMESTAMP '2016-10-20 11:15:25' group by trip_id
```
Then we run into OOM issues *if* we don't have `clearRACBuffers()` because we keep accumulating data in `resultRowAndCols`.

I added the comment to clarify the above only, let me know if I should work on elaborating the comment a bit?

> this method sets rowId to 0 - if that would happen when we do have something still to write; that will make it loose data...

This part at the beginning of `runIncrementally()` ensures that we didn't have any pending data:
```java
if (frameHasRowsPendingFlush()) {
      flushAllRowsAndCols();
      return ReturnOrAwait.runAgain();
    }
```
...so it's safe to clear the buffers at this point.
(Note that the buffers are getting cleared at the beginning of `non-empty over()` code flow)

> why can't the same be done when the last part of the frame was written?
call it at the end of writeRacToFrame ?

Because `writeRacToFrame` doesn't necessarily writes the whole RAC to the frame, it writes only upto the frame writer's capacity.

Reply:
> Because writeRacToFrame doesn't necessarily writes the whole RAC to the frame, it writes only upto the frame writer's capacity.

is the following true:
* there is an early return in that method; which is excercised if its not processed fully
* if its not processed fully the `frameHasRowsPendingFlush` will return true
* the end of that method is only reachable if the full input is processed

if yes - then why not call `clearRACBuffers` there and remove these comments?

Reply:
@kgyrtkirk My bad, yes you're right. Have made the change, thanks!